tsl/src/nodes/decompress_chunk/planner.c

```diff
@@ -40,169 +41,219 @@ _decompress_chunk_init(void)
-	FormData_hypertable_compression *ht_info =
-		get_column_compressioninfo(path->info->hypertable_compression_info, ht_attname);
-	AttrNumber scan_varattno = get_compressed_attno(path->info, ht_attno);
-	AttrNumber chunk_attno = get_attnum(path->info->chunk_rte->relid, ht_attname);
-
-	Assert(!get_rte_attribute_is_dropped(path->info->ht_rte, ht_attno));
-	Assert(!get_rte_attribute_is_dropped(path->info->chunk_rte, chunk_attno));
-	Assert(!get_rte_attribute_is_dropped(path->info->compressed_rte, scan_varattno));
-
-	if (ht_info->algo_id == _INVALID_COMPRESSION_ALGORITHM)
+	/*
+	 * We allow tableoid system column, it won't be in the targetlist but will
+	 * be added at decompression time. Always mark it as found.
+	 */
+	if (bms_is_member(TableOidAttributeNumber - FirstLowInvalidHeapAttributeNumber,
+					  chunk_attrs_needed))
 	{
-		Oid typid, collid;
-		int32 typmod;
-		get_atttypetypmodcoll(path->info->ht_rte->relid, ht_attno, &typid, &typmod, &collid);
-		scan_var =
-			makeVar(path->info->compressed_rel->relid, scan_varattno, typid, typmod, collid, 0);
+		chunk_attrs_found =
+			bms_add_member(chunk_attrs_found,
+						   TableOidAttributeNumber - FirstLowInvalidHeapAttributeNumber);
 	}
-	else
-		scan_var = makeVar(path->info->compressed_rel->relid,
-						   scan_varattno,
-						   ts_custom_type_cache_get(CUSTOM_TYPE_COMPRESSED_DATA)->type_oid,
-						   -1,
-						   0,
-						   0);
-	path->decompression_map = lappend_int(path->decompression_map, chunk_attno);
-
-	return makeTargetEntry((Expr *) scan_var, tle_index, NULL, false);
-}
-
-/*
- * build targetlist for scan on compressed chunk
- *
- * Since we do not adjust selectedCols in RangeTblEntry for chunks
- * we use selectedCols from the hypertable RangeTblEntry to
- * build the target list for the compressed chunk and adjust
- * attno accordingly
- */
-static List *
-build_scan_tlist(DecompressChunkPath *path)
-{
-	List *scan_tlist = NIL;
-	Bitmapset *attrs_used = path->info->ht_rte->selectedCols;
-	TargetEntry *tle;
-	int bit;
 
-	path->decompression_map = NIL;
+	/*
+	 * Fill the helper array of compressed attno -> compression info.
+	 */
+	FormData_hypertable_compression **compressed_attno_to_compression_info =
+		palloc0(sizeof(void *) * (path->info->compressed_rel->max_attr + 1));
+	ListCell *lc;
+	foreach (lc, path->info->hypertable_compression_info)
+	{
+		FormData_hypertable_compression *fd = lfirst(lc);
+		AttrNumber compressed_attno =
+			get_attnum(path->info->compressed_rte->relid, NameStr(fd->attname));
 
-	/* add count column */
-	tle = make_compressed_scan_meta_targetentry(path,
-												COMPRESSION_COLUMN_METADATA_COUNT_NAME,
-												DECOMPRESS_CHUNK_COUNT_ID,
-												list_length(scan_tlist) + 1);
-	scan_tlist = lappend(scan_tlist, tle);
+		if (compressed_attno == InvalidAttrNumber)
+		{
+			elog(ERROR,
+				 "column '%s' not found in the compressed chunk '%s'",
+				 NameStr(fd->attname),
+				 get_rel_name(path->info->compressed_rte->relid));
```

Comment:
Would be good to have a test for this case, just to make sure that it works as expected.

Reply:
Not sure how valuable this is, this would be the user tinkering with the compressed chunk table manually.

Reply:
It's a nit, so you decide how valuable you think it is, but imagine that `fd` or `path` happens be be NULL here. Then we have a crashing bug rather than an error message.

Reply:
I agree, recently had to fix a segfault in error handling not covered by the tests. But it was more complicated. Here we access both variables a couple of lines before, so I think this particular case is OK.

Reply:
Shouldn't we have at least a share lock on the RTEs?

Reply:
For compressed chunks we take the lock in `decompress_chunk_make_rte`. In general I think it is reasonable to assume for all RTEs we have in a query, otherwise it would be difficult to examine these tables in a concurrent environment.